### ISSUE TITLE
Proposal for safer `setState`

### ISSUE THREAD

**Issue by lukehutch** [flutter/flutter#147605]
### Use case

I frequently need to write code like:

```dart
  onTap: loadingProfileView
      ? null
      : () async {
          setState(() {
            loadingProfileView = true;
          });
          try {
            await openOtherUserProfileView(
              userId: profileThumbInfo.userId,
            );
          } finally {
            safeSetState(() {
              loadingProfileView = false;
            });
          }
        },
```

I have to use a safer version of `setState`, which I call `safeSetState`, after the async delay, because the widget may not be mounted, which causes an exception to be thrown.

### Proposal

Here is my code for `safeSetState`:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

extension SafeUpdateState on State {
  void safeSetState(void Function() updaterFunction) {
    void callSetState() {
      // Can only call setState if mounted
      if (mounted) {
        // ignore: invalid_use_of_protected_member
        setState(updaterFunction);
      }
    }

    if (SchedulerBinding.instance.schedulerPhase ==
        SchedulerPhase.persistentCallbacks) {
      // Currently building, can't call setState -- add post-frame callback
      SchedulerBinding.instance.addPostFrameCallback((_) => callSetState());
    } else {
      callSetState();
    }
  }
}
```

In general, it would be great if `setState` could be called from anywhere without negative consequences. (For this reason, I also made it possible to call `safeSetState` during `build`, even though I know that is bad practice.)

Could the behavior of the regular `setState` be updated to handle these cases, e.g. to simply fail silently if the widget is no longer mounted?

**Comment by lukehutch** [comment]
It's not really related to either of them: both of those issues were about how to solve a complex scheduling issue I had when I was trying to create a new type of reactive widget. The latter of those issues was probably caused by calling `SchedulerBinding.instance.addPostFrameCallback` without first checking `SchedulerBinding.instance.schedulerPhase` (which is fixed in my new `safeSetState`).

This issue is much more general: I am trying to solve the red screen of death issue that I get when I accidentally call `setState` on a widget that is no longer mounted, after an `await`. I believe that that should never be an app-stopping error -- if a widget is no longer mounted, at a minimum `setState` should just silently fail.

The extra scheduler phase check is simply to allow `setState` to be scheduled to be run even during `build` (which would generally be a bad idea, because `build` should have no side effects). But the reason for doing this is so that `safeSetState` can be called from any function, with the widget tree in any state.

Basically I am proposing that `setState` should never throw an exception for any reason.

**Comment by darshankawar** [comment]
Thanks for the update.
